src/tags/cycle_tag.rs

```diff
@@ -0,0 +1,183 @@
+use Renderable;
+use context::Context;
+use LiquidOptions;
+use token::Token::{self, Comma, Colon};
+use error::{Error, Result};
+use parser::{consume_value_token, value_token};
+
+use std::hash::{Hash, Hasher, SipHasher};
+
+
+struct Cycle {
+    name: String,
+    values: Vec<Token>
+}
+
+impl Renderable for Cycle {
+    // add code here
+    fn render(&self, context: &mut Context) -> Result<Option<String>> {
+        let value = try!(context.cycle_element(&self.name, &self.values));
+        Ok(value.map(|v| v.to_string()))
+    }
+}
+
+/// Internal implementation of cycle, to allow easier testing.
+fn parse_cycle(arguments: &[Token],
+               _options: &LiquidOptions) -> Result<Cycle> {
+    let mut args = arguments.iter();
+    let mut values = Vec::new();
+    let mut name = String::new();
+    let first = try!(consume_value_token(&mut args));
+
+    match args.next() {
+        Some(&Colon) => {
+            // the first argument is the name of the cycle block
+            name = first.to_string();
+        },
+        Some(&Comma) | None => {
+            // first argument is the first item in the cycle
+            values.push(first);
+        },
+        x => {
+            return Error::parser(": | Number | String | Identifier", x)
+        }
+    }
+
+    loop {
+        match args.next() {
+            Some(a) => {
+                let v = try!(value_token(a.clone()));
+                values.push(v);
+            },
+            None => break
+        }
+
+        match args.next() {
+            Some(&Comma) => {},
+            None => break,
+            x => return Error::parser("Comma", x)
+        }
+    }
+
+    if name.is_empty() {
+        let mut hasher = SipHasher::new();
```

Comment:
Not really - I just wanted a way to shorten the name and still have it unique. It may be premature optimisation, though.